src/daemon/rpmostreed-deployment-utils.cxx

```diff
@@ -103,6 +105,77 @@ rpmostreed_deployment_get_for_index (OstreeSysroot *sysroot,
   return (OstreeDeployment*)g_object_ref (deployments->pdata[deployment_index]);
 }
 
+static char*
+get_path_for_cached_signatures_variant (const gchar *remote,
+                                        const gchar *checksum)
+{
+  if (remote == NULL)
+    return g_build_filename (RPM_OSTREED_COMMIT_VERIFICATION_CACHE, "local", checksum, NULL);
+  return g_build_filename (RPM_OSTREED_COMMIT_VERIFICATION_CACHE, "remote", remote, checksum, NULL);
+}
+
+static GVariant*
+get_cached_signatures_variant (OstreeRepo  *repo,
+                               const gchar *remote,
+                               const gchar *checksum,
+                               GError     **error)
+{
+  gboolean cache_hit = FALSE;
+
+  struct stat stbuf;
+  g_autofree char *path = get_path_for_cached_signatures_variant (remote, checksum);
+  if (!glnx_fstatat_allow_noent (AT_FDCWD, path, &stbuf, AT_SYMLINK_NOFOLLOW, error))
+    return (GVariant*)glnx_prefix_error_null (error, "querying cached signature variant");
+  else if (errno == 0)
+    {
+      if (stbuf.st_mode == (S_IFREG | 0600) && stbuf.st_uid == 0 && stbuf.st_gid == 0)
+        cache_hit = TRUE;
+    }
+  else if (errno != ENOENT)
+    {
+      g_assert_not_reached ();
+    }
+
+  if (!cache_hit)
+    {
+      if (!glnx_shutil_mkdir_p_at (AT_FDCWD, dirname (strdupa (path)), 0700, NULL, error))
+        return NULL;
+      if (TEMP_FAILURE_RETRY (unlink (path)) && errno != ENOENT)
+        return (GVariant*)glnx_null_throw_errno_prefix (error, "unlink(%s)", path);
+
+      g_autoptr(OstreeGpgVerifyResult) verify_result =
+        ostree_repo_verify_commit_for_remote (repo, checksum, remote, NULL, error);
+      if (!verify_result)
+        return NULL;
+
+      g_auto(GVariantBuilder) builder;
+      g_variant_builder_init (&builder, G_VARIANT_TYPE ("av"));
+      guint n_sigs = ostree_gpg_verify_result_count_all (verify_result);
```

Comment:
AFAICS, all this stuff is bound in Rust, e.g. https://docs.rs/ostree/0.13.5/ostree/struct.Repo.html#method.verify_commit_for_remote

Did you look at that?  It'd be nice to have new security-relevant code in Rust.

That said this looks generally sane to me.

Reply:
Heh, I had a feeling this would come up. It started small and each incremental bit was easier to add in C. I can RIIR though might not circle back to it right away.